pkg/controller/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -0,0 +1,1227 @@
+			message,
+			controllerutils.UpdateConditionIfReasonOrMessageChange)
+	}
+
+	var readyChanged bool
+	if ready := controllerutils.FindClusterDeploymentCondition(conditions, hivev1.AWSPrivateLinkReadyClusterDeploymentCondition); ready == nil ||
+		ready.Status != corev1.ConditionTrue {
+		conditions, readyChanged = controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+			conditions,
+			hivev1.AWSPrivateLinkReadyClusterDeploymentCondition,
+			completed,
+			reason,
+			message,
+			controllerutils.UpdateConditionIfReasonOrMessageChange)
+	}
+	if !readyChanged && !failedChanged {
+		return nil
+	}
+	curr.Status.Conditions = conditions
+	logger.Debug("setting AWSPrivateLinkReadyClusterDeploymentCondition to %s", completed)
+	return r.Status().Update(context.TODO(), curr)
+}
+
+func (r *ReconcileAWSPrivateLink) reconcilePrivateLink(cd *hivev1.ClusterDeployment, clusterMetadata *hivev1.ClusterMetadata, logger log.FieldLogger) (reconcile.Result, error) {
+	logger.Debug("reconciling PrivateLink resources")
+	awsClient, err := newAWSClient(r,
+		corev1.SecretReference{Name: cd.Spec.Platform.AWS.CredentialsSecretRef.Name, Namespace: cd.Namespace},
+		cd.Spec.Platform.AWS.Region)
+	if err != nil {
+		logger.WithError(err).Error("error creating AWS client for the cluster")
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setProgressCondition(cd, corev1.ConditionFalse,
+		"DiscoveringNLB",
+		"discovering NLB for the cluster",
+		logger); err != nil {
+		logger.WithError(err).Error("failed to update condition on cluster deployment")
+		return reconcile.Result{}, err
+	}
+
+	// discover the NLB for the cluster.
+	nlbARN, err := discoverNLBForCluster(awsClient.user, clusterMetadata.InfraID, logger)
+	if err != nil {
+		if awsErrCodeEquals(err, "LoadBalancerNotFound") {
+			logger.WithField("infraID", clusterMetadata.InfraID).Debug("NLB is not yet created for the cluster, will retry later")
+
+			if err := r.setProgressCondition(cd, corev1.ConditionFalse,
+				"DiscoveringNLBNotYetFound",
+				"discovering NLB for the cluster, but it does not exists yet",
+				logger); err != nil {
+				logger.WithError(err).Error("failed to update condition on cluster deployment")
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{RequeueAfter: defaultRequeueLater}, nil
+		}
+
+		logger.WithField("infraID", clusterMetadata.InfraID).WithError(err).Error("error discovering NLB for the cluster")
+
+		if err := r.setErrCondition(cd, "DiscoveringNLBFailed", err, logger); err != nil {
+			logger.WithError(err).Error("failed to update condition on cluster deployment")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setProgressCondition(cd, corev1.ConditionFalse,
+		"ReconcilingVPCEndpointService",
+		"reconciling the VPC Endpoint Service for the cluster",
+		logger); err != nil {
+		logger.WithError(err).Error("failed to update condition on cluster deployment")
+		return reconcile.Result{}, err
+	}
```

Comment:
Similar here, if we drop this set progress, and instead in the block below we track if we had to do anything or not as a change bool from reconcileVPCEndpointService, and if it's true we set Ready progress to ReconciledVPCEndpointService. I think essentially I'm proposing the condition reasons change from "what i'm about to do" to " what I last did", this will cut down on the etcd writes in the event of requeues, and seemingly going back in time to steps we previously did successfully if you're watching the condition.

Granted if we start getting failures it won't be as obvious where in the Ready condition which would now show us the last thing we did but not the thing we're doing currently, but it will be visible in the Failure condition, so in tandem it should be quite clear where the issue is surfacing.

Same for each Ready progress update below.

Reply:
i'm hoping 
43e3e78  is what you had in mind @dgoodwin

Reply:
Yes this should do the trick, thanks.

You have some conflicts, otherwise LGTM, squash anything you want to.